lib/nostrum/shard/session.ex

```diff
@@ -81,59 +110,76 @@ defmodule Nostrum.Shard.Session do
 
     case from_handle do
       {new_state, reply} ->
-        {:reply, {:binary, reply}, new_state}
+        :ok = :gun.ws_send(state.conn, {:binary, reply})
+        {:noreply, new_state}
 
       new_state ->
-        {:ok, new_state}
+        {:noreply, new_state}
     end
   end
 
-  def handle_cast({:status_update, payload}, state) do
-    {:reply, {:binary, payload}, state}
-  end
+  def handle_info(
+        {:gun_down, conn, protocol, reason, _killed_streams, _unprocessed_streams},
+        state
+      ) do
+    Logger.warn(fn -> "connection down (#{protocol} #{reason}), attempting reconnect" end)
 
-  def handle_cast({:request_guild_members, payload}, state) do
-    {:reply, {:binary, payload}, state}
-  end
-
-  def handle_cast(:heartbeat, %{heartbeat_ack: false} = state) do
-    Logger.warn("heartbeat_ack not received in time, disconnecting")
-    {:close, state}
-  end
+    {:ok, :cancel} = :timer.cancel(state.heartbeat_ref)
+    :ok = :gun.close(conn)
 
-  def handle_cast(:heartbeat, state) do
-    {:ok, ref} =
-      :timer.apply_after(state.heartbeat_interval, WebSockex, :cast, [state.conn_pid, :heartbeat])
+    worker = connect(state.gateway)
 
-    {:reply, {:binary, Payload.heartbeat_payload(state.seq)},
-     %{state | heartbeat_ref: ref, heartbeat_ack: false, last_heartbeat_send: DateTime.utc_now()}}
+    {:noreply, %{state | conn: worker}}
   end
 
-  def handle_disconnect(%{reason: reason}, state) when is_tuple(reason) do
+  def handle_info({:gun_ws, conn, _stream, {:close, errno, reason}}, state) do
     Logger.warn(fn ->
-      "websocket disconnected with reason #{inspect(reason)}, attempting reconnect"
+      "websocket disconnected with code #{errno}, reason #{inspect(reason)}, " <>
+        "attempting reconnect"
     end)
 
-    :timer.cancel(state.heartbeat_ref)
+    {:ok, :cancel} = :timer.cancel(state.heartbeat_ref)
+    :ok = :gun.close(conn)
 
-    {:reconnect, state}
+    # Drop the `:gun_down` message from the mailbox.
+    receive do
```

Comment:
We shouldn't do this here. If we're inside of a callback I don't think there's any guarantee we'll get sent a message by gun here and it could be causing the deadlock issue you were talking about. We should instead add another `handle_info` for a gun_down message and let it come in naturally.  

For both of these we can return `:noreply` (and maybe a helpful logger error message) at which point the ws connection should close and gun will try to reconnect. Gun will send a `gun_up` message as stated [here](https://ninenines.eu/docs/en/gun/1.0/manual/gun_up/) if it reconnects successfully and then we can upgrade to ws again and send a resume to Discord.